File: src/python/problems/close_strings_test.py
# https://leetcode.com/problems/determine-if-two-strings-are-close/

from core.problem_base import *
import collections

class CloseStrings(ProblemBase):
    def Solution(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False
        
        map1 = collections.Counter(word1)
        map2 = collections.Counter(word2)

        if len(map1) != len(map2):
            return False

        for key in map1.keys():
            if key not in map2:
                return False
        
        occ1 = sorted(map1.values())
        occ2 = sorted(map2.values())

        return occ1 == occ2


if __name__ == '__main__':
    TestGen(CloseStrings) \
        .Add(lambda tc: tc.Param("abc").Param("bca").Result(True)) \
        .Add(lambda tc: tc.Param("a").Param("a").Result(True)) \
        .Add(lambda tc: tc.Param("a").Param("aa").Result(False)) \
        .Add(lambda tc: tc.Param("cabbba").Param("abbccc").Result(True)) \
        .Add(lambda tc: tc.Param("abbzzca").Param("babzzcz").Result(False)) \
        .Run()

File: src/go/problems/close_strings_test.go
// https://leetcode.com/problems/determine-if-two-strings-are-close

package problems

import (
	"sort"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type CloseStrings struct{}

func TestCloseStrings(t *testing.T) {
	gen := core.TestSuite[CloseStrings]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("abc").Param("bca").Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param("a").Param("aa").Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param("cabbba").Param("abbccc").Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param("abbzzca").Param("babzzcz").Result(false)
	}).Run(t)
}

func (CloseStrings) Solution(word1 string, word2 string) bool {
	if len(word1) != len(word2) {
		return false
	}

	count := func(word string) []int {
		set := make([]int, 26)
		for i := 0; i < len(word); i++ {
			set[word[i]-'a']++
		}
		return set
	}

	map1 := count(word1)
	map2 := count(word2)

	for i := 0; i < len(map1); i++ {
		if (map1[i] == 0 && map2[i] > 0) || (map1[i] > 0 && map2[i] == 0) {
			return false
		}
	}

	sort.Ints(map1)
	sort.Ints(map2)

	for i := 0; i < len(map1); i++ {
		if map1[i] != map2[i] {
			return false
		}
	}

	return true
}
